### PR TITLE
feat: add list_consumable command

### PR DESCRIPTION
**Pull request by tomyrd**
Add a `list_consumable` command that lists all consumable notes. It can receive an optional `account_id` argument so that it only shows notes that can be consumed by that account.

### CODE REVIEW

File: src/store/note_record/input_note_record.rs [hunk 1]
```diff
@@ -187,3 +187,31 @@ impl TryInto<InputNote> for InputNoteRecord {
         }
     }
 }
+
+impl TryInto<Note> for InputNoteRecord {
+    type Error = ClientError;
+
+    fn try_into(self) -> Result<Note, Self::Error> {
```
[the remark this answers is not in the document]
Comment (mFragaBA) [reply src/store/note_record/input_note_record.rs:194]: Agreed! You won't probably need the inclusion proof to build a `Note`, and yes you'll need to cover situations where we only know part of the note's Info. You can read more on [this in this comment that bobbin wrote](https://github.com/0xPolygonMiden/miden-client/issues/98#issuecomment-1905254363) and subsequent comments

File: src/cli/input_notes.rs [hunk 2]
```diff
@@ -280,6 +302,30 @@ where
...
+fn print_consumable_notes_summary<'a, I>(notes: I) -> Result<(), String>
+where
+    I: IntoIterator<Item = &'a ConsumableNote>,
+{
+    let mut table = create_dynamic_table(&["Note ID", "Account ID", "Relevance"]);
+
+    for note in notes {
+        for relevance in &note.1 {
+            table.add_row(vec![
+                note.0.id().inner().to_string(),
```
Comment (mFragaBA) [inline src/cli/input_notes.rs:314]: I think you can use `id().to_hex()` here

File: src/cli/input_notes.rs [hunk 3]
```diff
@@ -280,6 +302,30 @@ where
...
+                relevance.0.to_string(),
+                match relevance.1 {
+                    NoteRelevance::Always => "Always".to_string(),
+                    NoteRelevance::After(height) => format!("After block {}", height),
+                },
```
Comment (mFragaBA) [inline src/cli/input_notes.rs:319]: nit: we should implement `Display` on `NoteRelevance` with a match similar to this one and here just use `relevance.1.to_string()`

**Review by mFragaBA — CHANGES_REQUESTED** [review]

File: src/store/note_record/input_note_record.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (tomyrd) [reply src/store/note_record/input_note_record.rs:194]: Thanks for letting me know. I changed the `TryInto` implementation to only check for the metadata. From what I read from the comment, the rest of the fields used (inputs, serial_num, script and assets) are all known for input notes. So i don't think it's necessary to cover more cases. Am i correct?

File: src/client/notes.rs [hunk 4]
```diff
@@ -15,6 +20,40 @@ impl<N: NodeRpcClient, R: FeltRng, S: Store> Client<N, R, S> {
...
+        account_id: &Option<String>,
+    ) -> Result<Vec<ConsumableNote>, ClientError> {
+        let commited_notes = self.store.get_input_notes(NoteFilter::Committed)?;
+
+        let note_screener = NoteScreener::new(&self.store);
+        let mut relevant_notes: Vec<ConsumableNote> = Vec::new();
+
+        for input_note in commited_notes {
+            let account_relevance =
+                note_screener.check_relevance(&input_note.clone().try_into().unwrap())?;
```
[the remark this answers is not in the document]
Comment (tomyrd) [reply src/client/notes.rs]: Solved. I also deleted other unnecessary unwraps in my code.

File: src/client/notes.rs [hunk 5]
```diff
@@ -15,6 +20,39 @@ impl<N: NodeRpcClient, R: FeltRng, S: Store> Client<N, R, S> {
...
+                if account_relevance.iter().any(|(id, _)| *id == account_id) {
+                    relevant_notes.push((
+                        input_note,
+                        account_relevance.into_iter().filter(|(id, _)| *id == account_id).collect(),
+                    ));
+                }
+            } else {
+                relevant_notes.push((input_note, account_relevance));
+            }
+        }
```
Comment (mFragaBA) [inline src/client/notes.rs:51]: What do you think about this? The idea is to split the filter by Id from the creation of the `ConsumableNote`s. It removes more nested Ifs, although I'm doubting whether the end result is more understandable or not.

```rust
        let mut relevant_notes = Vec::new();
        for input_note in commited_notes {
            let account_relevance =
                note_screener.check_relevance(&input_note.clone().try_into()?)?;

            if account_relevance.is_empty() { continue; }

            relevant_notes.push((input_note, account_relevance));
        }

        // If no account id was provided we already have the result
        if account_id.is_none() { return Ok(relevant_notes); }
        let account_id = account_id.expect("Should retrieve account id");

        // Use the given id to filter notes
        let relevant_notes = relevant_notes
            .into_iter()
            .filter(|(_input_note, relevance)| relevance.iter().any(|(id, _)| *id == account_id))
            .collect();

        Ok(relevant_notes)
```

File: src/store/note_record/input_note_record.rs [hunk 6]
```diff
@@ -187,3 +187,22 @@ impl TryInto<InputNote> for InputNoteRecord {
...
+            Some(metadata) => {
+                let note_inputs = NoteInputs::new(self.details.inputs)?;
+                let note_recipient =
+                    NoteRecipient::new(self.details.serial_num, self.details.script, note_inputs);
+                let note = Note::new(self.assets, metadata, note_recipient);
+                Ok(note)
+            },
+            None => Err(ClientError::NoteError(miden_objects::NoteError::invalid_origin_index(
+                "Input Note Record contains no metadata".to_string(),
+            ))),
```
Comment (mFragaBA) [inline src/store/note_record/input_note_record.rs:205]: This works for now but we'll probably need to add a `NoteRecordError` to `ClientError` in the future.

**Review by mFragaBA — COMMENTED** [review]
Can we add tests for `get_consumable_notes`? Either Integration/unit tests are fine (probably integration tests will be easier due to not needing to mock a lot of stuff but you can choose whichever is more comfortable to you)

File: src/client/notes.rs [hunk 7]
```diff
@@ -1,11 +1,16 @@
...
+use super::{note_screener::NoteRelevance, rpc::NodeRpcClient, Client};
 use crate::{
+    client::NoteScreener,
     errors::ClientError,
     store::{InputNoteRecord, NoteFilter, Store},
 };
 
+// TYPES
+// --------------------------------------------------------------------------------------------
+pub type ConsumableNote = (InputNoteRecord, Vec<(AccountId, NoteRelevance)>);
```
Comment (mFragaBA) [inline src/client/notes.rs:12]: If this is public we should add at least some minimal documentation

File: src/store/note_record/input_note_record.rs (hunk 6 above)
Comment (igamigo) [reply src/store/note_record/input_note_record.rs:205]: Yes, let's do this if possible. Let's add a `ClientError::NoteRecordError(String)` that explains that the note does not contain metadata and the conversion is impossible.

File: src/store/note_record/input_note_record.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (igamigo) [reply src/store/note_record/input_note_record.rs:194]: Yeah, that looks correct!

**Review by mFragaBA — APPROVED** [review]
LGTM!

**Pull request closed without merging**